solr/core/src/java/org/apache/solr/update/processor/DistributedUpdateProcessor.java

```diff
@@ -663,118 +661,36 @@ boolean getUpdatedDocument(AddUpdateCommand cmd, long versionOnUpdate) throws IO
     final boolean isDeeplyNestedSchema = req.getSchema().isUsableForChildDocs() && req.getSchema().hasExplicitField(IndexSchema.NEST_PATH_FIELD_NAME);
     SolrInputDocument sdoc = cmd.getSolrInputDocument();
     BytesRef id = cmd.getIndexedId();
-    SolrInputDocument completeHierarchy = RealTimeGetComponent.getInputDocument(cmd.getReq().getCore(), id, RealTimeGetComponent.Resolution.ROOT_WITH_CHILDREN);
+    SolrInputDocument oldRootDocWithChildren = RealTimeGetComponent.getInputDocument(cmd.getReq().getCore(), id, RealTimeGetComponent.Resolution.ROOT_WITH_CHILDREN);
 
-    if (completeHierarchy == null) {
+    if (oldRootDocWithChildren == null) {
       if (versionOnUpdate > 0) {
         // could just let the optimistic locking throw the error
         throw new SolrException(ErrorCode.CONFLICT, "Document not found for update.  id=" + cmd.getPrintableId());
       }
     } else {
-      completeHierarchy.remove(CommonParams.VERSION_FIELD);
+      oldRootDocWithChildren.remove(CommonParams.VERSION_FIELD);
     }
 
 
     SolrInputDocument mergedDoc;
-    if(idField == null || completeHierarchy == null) {
+    if(idField == null || oldRootDocWithChildren == null) {
       // create a new doc by default if an old one wasn't found
       mergedDoc = docMerger.merge(sdoc, new SolrInputDocument());
     } else {
-      if(isDeeplyNestedSchema && !isRootDoc(sdoc, completeHierarchy)) {
-        // this is an update and the updated doc is not the root document
+      if(isDeeplyNestedSchema && !docMerger.isRootDoc(sdoc, oldRootDocWithChildren)) {
+        // this is an update where the updated doc is not the root document
         SolrInputDocument sdocWithChildren = RealTimeGetComponent.getInputDocument(cmd.getReq().getCore(),
             id, RealTimeGetComponent.Resolution.DOC_WITH_CHILDREN);
-        // get path of document to be updated
-        String updatedDocPath = (String) sdocWithChildren.getFieldValue(IndexSchema.NEST_PATH_FIELD_NAME);
-        // get the SolrInputField containing the document which the AddUpdateCommand updates
-        SolrInputField sifToReplace = getFieldFromHierarchy(completeHierarchy, updatedDocPath);
-        // update SolrInputField, either appending or replacing the updated document
-        updateDocInSif(sifToReplace, sdocWithChildren, sdoc);
-        mergedDoc = completeHierarchy;
+        mergedDoc = docMerger.mergeNonRoot(sdoc, oldRootDocWithChildren, sdocWithChildren);
```

Comment:
`mergeChildDoc` would probably be clearer than `mergeNonRoot`?  It's usually clearer to avoid negatives when we can say clearly what it _is_ versus what it _is not_.

Reply:
Sure thing.